Assignment 4/src/FractionalWrapper.sol

```diff
@@ -0,0 +1,302 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import "yield-utils-v2/access/Ownable.sol";
+import "yield-utils-v2/token/ERC20.sol";
+import "yield-utils-v2/token/TransferHelper.sol";
+import "src/EIP4626.sol";
+
+error TransferFailed();
+error DepositFailed();
+error WithdrawalFailed();
+
+using TransferHelper for IERC20;
+
+/// @title A fractional wrapper.
+/// @notice Wraps a token in a new token with a different denomination.
+///     Implements ERC20 as well as the vault standard EIP4626.
+/// @author Thomas den Hollander
+/// @dev The fraction is set in the constructor. Apart from the `convert*`
+///     functions that do the fractional computation, the vault behaviour is
+///     pretty standard.
+contract FractionalWrapper is
+    EIP4626,
+    ERC20("FractionalWrapper", "FWRAP", 18),
+    Ownable
+{
+    /// @notice The asset that is used as underlying in this contract.
+    IERC20 public immutable token;
```

Comment:
You can just use `asset`.

This:
`address public immutable override asset;`

Overrides this:
`function asset() external view override returns (address);`

You could go even further and modify the EIP4626.sol that you inherit from to make `asset` an IERC20. You need to conform with the standard at an ABI level, and at a functionality level. Declaring a variable as a contract or an address is only visible to the compiler, so won't make you lose conformity.

Reply:
That's good to know, thanks!